File: src/main/java/chapter5/example/CallbackHeaven.java
package chapter5.example;

import common.CommonUtils;
import common.Log;
import common.OkHttpHelper;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;


public class CallbackHeaven {

    private static final String FIRST_URL = "https://api.github.com/zen";
    private static final String SECOND_URL = CommonUtils.GITHUB_ROOT + "/samples/callback_heaven";

    public void usingConcat() {
        CommonUtils.exampleStart();
        Observable<String> source = Observable.just(FIRST_URL)
                .subscribeOn(Schedulers.io())
                .map(OkHttpHelper::get)
                .concatWith(Observable.just(SECOND_URL)
                        .map(OkHttpHelper::get));
        source.subscribe(Log::it);
        CommonUtils.sleep(5000);
    }

    public void usingZip() {
        CommonUtils.exampleStart();
        Observable<String> first = Observable.just(FIRST_URL)
                .subscribeOn(Schedulers.io())
                .map(OkHttpHelper::get);
        Observable<String> second = Observable.just(SECOND_URL)
                .subscribeOn(Schedulers.io())
                .map(OkHttpHelper::get);

        Observable.zip(first, second, (a, b) -> ("\n>> " + a + "\n>> " + b))
                .subscribe(Log::it);

        CommonUtils.sleep(5000);
    }

    public static void main(String[] args) {
        CallbackHeaven demo = new CallbackHeaven();
        // demo.usingConcat();
        demo.usingZip();
    }

}

File: src/main/kotlin/chapter5/example/CallbackHeaven_.kt
package chapter5.example

import common.CommonUtils_
import common.Log_
import common.OkHttpHelper_
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers


class CallbackHeaven_ {

    private val FIRST_URL = "https://api.github.com/zen"
    private val SECOND_URL = CommonUtils_.GITHUB_ROOT + "/samples/callback_heaven"


    fun usingConcat() {
        CommonUtils_.exampleStart()
        val source = Observable.just(FIRST_URL)
                .subscribeOn(Schedulers.io())
                .map { OkHttpHelper_.get(it) }
                .concatWith(Observable.just(SECOND_URL)
                        .map { OkHttpHelper_.get(it) })
        source.subscribe { Log_.it(it) }
        CommonUtils_.sleep(5000)
    }

    fun usingZip() {
        CommonUtils_.exampleStart()
        val first = Observable.just(FIRST_URL)
                .subscribeOn(Schedulers.io())
                .map { OkHttpHelper_.get(it) }
        val second = Observable.just(SECOND_URL)
                .subscribeOn(Schedulers.io())
                .map { OkHttpHelper_.get(it) }

        Observable.zip(first, second, BiFunction { a: String, b: String -> "\n>> $a\n>> $b" })
                .subscribe { Log_.it(it) }

        CommonUtils_.sleep(5000)
    }

}


fun main(args: Array<String>) {
    val demo = CallbackHeaven_()
    // demo.usingConcat()
    demo.usingZip()
}
